AI assistant: Past sessions list

jurgenwerk opened the pull request:
Opens past AI sessions list in a popover. Previously it was shown as a simple list in the panel.

https://github.com/cardstack/boxel/assets/273660/4f028f11-ed53-4d6b-add8-6d081d5eb653

jurgenwerk committed:
Show past sessions list in a popover

jurgenwerk committed:
Layout fixes

jurgenwerk commented on packages/host/app/components/ai-assistant/panel-popover.gts:
```diff
@@ -0,0 +1,45 @@
+  };
+}
+
+export default class AiAssistantPanelPopover extends Component<Signature> {
```
This component is made to be reused when we add other popovers such as "Rename session":

<img width="456" alt="image" src="https://github.com/cardstack/boxel/assets/273660/882e1cf9-5878-4cdc-95d3-839a60b49d94">

jurgenwerk committed:
Move test selector to the appropriate place

jurgenwerk commented on packages/base/room.gts:
```diff
@@ -369,7 +369,7 @@ export class RoomField extends FieldDef {
           displayName: event.content.displayname,
           membership: event.content.membership,
-          membershipTs: event.origin_server_ts,
+          membershipTs: event.origin_server_ts || Date.now(),
```
I added this to show the session date in tests

jurgenwerk commented:
I'm working on fixing matrix tests.

jurgenwerk committed:
Fix rooms/sessions related tests now that they are in a popover

jurgenwerk committed:
Fix a mistake in how panel height is calculated

jurgenwerk committed:
Merge branch 'main' into past-sessions-list

jurgenwerk commented:
So the issue with failing tests is that *sometimes* this happens (some rooms have a blank name while in reality the name is present), and the tests that that include asserting the items in room list will fail:

<img width="415" alt="image" src="https://github.com/cardstack/boxel/assets/273660/2e883f2b-6c2e-4a8f-8c3c-15f37aa93296">

This is painful to reproduce because in development it happens super rarely, and in testing it happens only sometimes.

My initial thought was that the matrix room event that tells us the room name is either not coming or not registered, but with debugging I ruled that out. The whole nature of the problem sounds like a race condition and that led me to discover one clue which is that the event that sets the room name in the case where room name is missing is registered **after** the `@cached private get roomResources()` is called... which makes me think the problem lies in improper caching or reactivity issues with room resources. We have multiple cached room/sessions getters in [ai-assistant/panel.gts](https://github.com/cardstack/boxel/blob/50756d10e861d64b1d4faa91c7cea79d30932007/packages/host/app/components/ai-assistant/panel.gts#L273) so I'm gonna check if perhaps refactoring to get rid of `@cached` solves the issue

jurgenwerk commented:
Illustration of the problem: 

<img width="1180" alt="image" src="https://github.com/cardstack/boxel/assets/273660/b19e6e9f-d7eb-4778-984c-74652266751b">

jurgenwerk committed:
Test fix attempt (works in dev)

jurgenwerk committed:
Cosmetic fix

jurgenwerk committed:
Not needed

jurgenwerk commented:
I don't fully understand the bug yet and why it is so difficult to reproduce it locally, but here is a simple workaround for the mentioned race condition (? - or some kind of a reactivity issue with caching in `newEvents` ): https://github.com/cardstack/boxel/pull/969/files#diff-f3a5500e52102cde45a7c53d626862d10b7b8789267400fcbdbb8389f292cb61R295

backspace commented:
I’ll look at the code after this but a minor improvement suggestion I have is to make the border width consistent here so hovering doesn’t cause a 1px shifting:

![screencast 2024-01-22 13-42-40](https://github.com/cardstack/boxel/assets/43280/aca62891-353a-4582-840c-c0fd23331944)

backspace commented on packages/host/app/components/ai-assistant/past-sessions.gts:
```diff
@@ -0,0 +1,80 @@
+
+      .bottom {
+        margin-top: var(--boxel-sp-4xs);
+        color: #7e7c8d;
```
Would `--boxel-450` suffice here? It looks close to me

jurgenwerk replied:
Yeah it's good!

<img width="255" alt="image" src="https://github.com/cardstack/boxel/assets/273660/25ef869d-33d0-4432-b47e-a2a8cdd71918">

backspace reviewed (approved):
I tried this out locally, looks good apart from the previously-commented suggestion 🤩

burieberry reviewed (commented):
Looks good. Maybe something like using transparent border-color instead of removing the border on hover can help with the shifting that buck mentioned

jurgenwerk committed:
Improve layout

jurgenwerk committed:
Empty collection indicator

jurgenwerk commented:
@burieberry cool yeah that worked!

jurgenwerk commented:
I addressed the comments about the layout, added fixes for that. Regarding the workaround for the bug mentioned where I read events from `events` instead of `newEvents`–this obviously comes with a performance hit but given we don't expect so many events, it shouldn't be noticeable, I think. @habdelra would love to have a chat with you about this when you're back
